crates/iota-grpc-api/src/types.rs

```diff
@@ -145,8 +145,31 @@ impl BcsData {
 // Type aliases and utility types
 pub type CheckpointStreamResult = Result<Checkpoint, Status>;
 
+// Storage abstraction traits for gRPC access
+// These traits provide an abstraction layer over the storage backend,
+// making it easier to implement gRPC services with different storage types
+// (e.g., production database vs simulacrum for testing).
+
+/// Trait for reading checkpoint data from storage
+pub trait GrpcStateReader: Send + Sync + 'static {
+    /// Get the latest checkpoint sequence number
+    fn get_latest_checkpoint_sequence(&self) -> Option<u64>;
+
+    /// Get checkpoint summary by sequence number
+    fn get_checkpoint_summary(&self, seq: u64) -> Option<CertifiedCheckpointSummary>;
+
+    /// Get full checkpoint data by sequence number
+    fn get_checkpoint_data(&self, seq: u64) -> Option<CheckpointData>;
+
+    /// Get epoch's last checkpoint for epoch boundary calculations
+    fn get_epoch_last_checkpoint(
+        &self,
+        epoch: u64,
+    ) -> Result<Option<CertifiedCheckpointSummary>, Box<dyn std::error::Error + Send + Sync>>;
```

Comment:
Why not use `anyhow::Error` or even `anyhow::Result` here?

Reply:
Thanks, fixed it at [9e4c5a2](https://github.com/iotaledger/iota/pull/8019/commits/9e4c5a2fbf40de761d1ed5151f003516cececbd9).